### PR TITLE
Structured responses from transcoder

### PR DESCRIPTION
**Pull request by j0sh**
* [x] Structural support for returning values from the transcoder
* [x] Serialize returned value over protobufs, send response
* [x] Fill in transcoder response values (URL, hash, sig, etc)
* [x] Implement response handling on broadcaster
* [x] Monitoring updates
* [x] Download segments from response
* [x] Update playlists
* [ ] Check response signature from transcoder
* [x] Move protobufs into `net` to facilitate access from `core`

### CODE REVIEW

**Comment by j0sh** [comment]
Note the structure of the data that we return to the client. We have a choice to make: do we need the hashes of the individual segments, or is it enough to just return the URLs and check the sig of the concatenated hashes? Currently, we return both the individual hashes and the concatenated hashes.

The current protocol also does not allow for this, but I wonder if we should make future allowances for 'partial' success in transcoding, eg only returning a subset of the requested profiles. This would make transcoding less all-or-nothing, and allow for some interesting constructs like 'streaming' responses. To support this, we would need to switch from `list[Segment]` to `map[StreamID]Segment` in that case.

https://github.com/livepeer/go-livepeer/blob/04152802ee86fda59de3af2b02e3dd62f0212b93/server/lp_rpc.proto#L69-L87

**Comment by j0sh** [comment]
> Note the structure of the data that we return to the client. We have a choice to make: do we need the hashes of the individual segments, or is it enough to just return the URLs and check the sig of the concatenated hashes? Currently, we return both the individual hashes and the concatenated hashes.

Reversed course here -- now we only return URLS and a signature. This was mostly out of convenience due to how I chose to return the concatenated hashes: from the claim manager, which already computes those hashes. https://github.com/livepeer/go-livepeer/pull/508/commits/4836db7f57ffd562d19faed2b94fc480080721a4 To return each individual hash, we would need to either refactor the claimmanager hashing procedure, or re-hash everything outside the claimmanager. Path of least resistance, but we can certainly change it up if folks think otherwise.

File: core/transcoder.go [hunk 1]
```diff
@@ -131,32 +146,30 @@ func (n *LivepeerNode) getSegmentChan(job *ethTypes.Job) (SegmentChan, error) {
...
+func (n *LivepeerNode) TranscodeSegment(job *ethTypes.Job, ss *SignedSegment) (*TranscodeResult, error) {
 	glog.V(common.DEBUG).Infof("Starting to transcode segment %v", ss.Seg.SeqNo)
 	ch, err := n.getSegmentChan(job)
 	if err != nil {
 		glog.Error("Could not find segment chan ", err)
-		return err
+		return nil, err
 	}
-	segChan := &SegChanData{seg: ss, res: make(chan error, 1)}
+	segChan := &SegChanData{seg: ss, res: make(chan *TranscodeResult, 1)}
```
Comment (ericxtang) [inline core/transcoder.go]: Small thing - `segChan` seems like a misnomer here.  Should be `segChanData`?

File: core/transcoder.go [hunk 2]
```diff
@@ -131,32 +146,30 @@ func (n *LivepeerNode) getSegmentChan(job *ethTypes.Job) (SegmentChan, error) {
...
-		}
+		return nil, fmt.Errorf("TranscoderBusy")
 	}
-	return nil
+	res := <-segChan.res
+	return res, res.Err
 }
 
-func (n *LivepeerNode) transcodeAndBroadcastSeg(seg *stream.HLSSegment, sig []byte, cm eth.ClaimManager, t transcoder.Transcoder, resultStrmIDs []StreamID, config net.TranscodeConfig) error {
+func (n *LivepeerNode) transcodeAndBroadcastSeg(config transcodeConfig, ss *SignedSegment) *TranscodeResult {
```
Comment (ericxtang) [inline core/transcoder.go]: Even though we are putting the transcoded video into the `VideoCache`, we are no longer "broadcasting" here.  Should we rename this to `transcodeAndCacheSeg`?

File: server/rpc.go [hunk 3]
```diff
@@ -177,7 +178,7 @@ func verifySegCreds(job *lpTypes.Job, segCreds string) (*SegData, error) {
 	return &segData, nil
 }
 
-func GetTranscoder(context context.Context, orch Orchestrator, req *TranscoderRequest) (*TranscoderInfo, error) {
+func GetTranscoder(context context.Context, orch Orchestrator, req *net.TranscoderRequest) (*net.TranscoderInfo, error) {
```
Comment (ericxtang) [inline server/rpc.go]: Seems `GetTranscoder` is only used in `lphttp.GetTranscoder`.  Make it a private method?

**Review by ericxtang — COMMENTED** [review]
Can we also remove the `net.VideoNetwork` from `VideoCache`?  That way it's just a pure cache, and not a weird "network-connected cache" anymore.

File: core/transcoder.go (hunk 1 above)
Comment (j0sh) [reply core/transcoder.go]: Fixed https://github.com/livepeer/go-livepeer/pull/512/commits/25e7e8b85b9038aee67a242f037d9794cf26d9d0

File: core/transcoder.go (hunk 2 above)
Comment (j0sh) [reply core/transcoder.go]: Fixed https://github.com/livepeer/go-livepeer/pull/512/commits/25e7e8b85b9038aee67a242f037d9794cf26d9d0

File: server/rpc.go (hunk 3 above)
Comment (j0sh) [reply server/rpc.go]: Fixed https://github.com/livepeer/go-livepeer/pull/512/commits/276c2c2ee4f2e58760cd458bf1670ab1cd32eb4a

**Comment by j0sh** [comment]
VideoNetwork was removed from VideoCache in https://github.com/livepeer/go-livepeer/pull/512/commits/2da29308070528f6f4c151292f4b264d26e0cfd3

**Review by yondonfu — APPROVED** [review]
🚢 Tracking the remaining todo here #523

**Pull request merged**